vdj_ann/src/transcript.rs

```diff
@@ -7,187 +7,252 @@ use crate::refx::RefData;
 use vector_utils::{lower_bound1_3, unique_sort};
 
+const MIN_DELTA: i32 = -25;
+const MIN_DELTA_IGH: i32 = -55;
+const MAX_DELTA: i32 = 35;
+
 // ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
 // TEST FOR VALID VDJ SEQUENCE
 // ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
 
-pub fn is_valid(
-    b: &DnaString,
-    refdata: &RefData,
-    ann: &[(i32, i32, i32, i32, i32)],
-    logme: bool,
-    log: &mut Vec<u8>,
-    is_gd: Option<bool>,
-) -> bool {
-    // Unwrap gamma/delta mode flag
-    let gd_mode = is_gd.unwrap_or(false);
-    let refs = &refdata.refs;
-    let rheaders = &refdata.rheaders;
-    for pass in 0..2 {
-        let mut m = "A";
-        if pass == 1 {
-            m = "B";
-        }
-        let mut vstarts = Vec::<i32>::new();
-        let mut jstops = Vec::<i32>::new();
-        let mut first_vstart: i32 = -1;
-        let mut first_vstart_len: i32 = -1;
-        let mut last_jstop: i32 = -1;
-        let mut last_jstop_len: i32 = -1;
-        let mut igh = false;
-        for j in 0..ann.len() {
-            let l = ann[j].0 as usize;
-            let len = ann[j].1 as usize;
-            let t = ann[j].2 as usize;
-            let p = ann[j].3 as usize;
-            if rheaders[t].contains("IGH") {
-                igh = true;
-            }
-            if !rheaders[t].contains("5'UTR")
-                && ((m == "A"
-                    && (rheaders[t].contains("TRAV")
-                        || (rheaders[t].contains("TRGV") && gd_mode)
-                        || rheaders[t].contains("IGHV")))
-                    || (m == "B"
-                        && (rheaders[t].contains("TRBV")
-                            || (rheaders[t].contains("TRDV") && gd_mode)
-                            || rheaders[t].contains("IGLV")
-                            || rheaders[t].contains("IGKV"))))
-            {
-                if first_vstart < 0 {
-                    first_vstart = l as i32;
-                    first_vstart_len = (refs[t].len() - p) as i32;
-                }
-                if p == 0 {
-                    vstarts.push(l as i32);
-                }
-            }
-            if (m == "A"
-                && (rheaders[t].contains("TRAJ")
-                    || (rheaders[t].contains("TRGJ") && gd_mode)
-                    || rheaders[t].contains("IGHJ")))
-                || (m == "B"
-                    && (rheaders[t].contains("TRBJ")
-                        || (rheaders[t].contains("TRDJ") && gd_mode)
-                        || rheaders[t].contains("IGLJ")
-                        || rheaders[t].contains("IGKJ")))
-            {
-                last_jstop = (l + len) as i32;
-                last_jstop_len = (p + len) as i32;
-                if p + len == refs[t].len() {
-                    jstops.push((l + len) as i32);
-                }
-            }
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Default)]
```

Comment:
We should only derive `Default` for types for which the default is actually a valid value.  Reading through the current implementation, we never return an instance of this type for which `full_length` or `has_v_start` is None, so these shouldn't be optional.  I'd suggest the following:
- Determine which fields are actually optional.
- Create a constructor that requires these fields to construct the type.
- Populate other fields via mutation as you do now.

Reply:
ContigStatus::Default is returned if a contig has no annotations or only v or only j-gene annotations.